_posts/2020-04-17-quarkus-mutual-tls.adoc

```diff
@@ -0,0 +1,405 @@
+
+keytool -genkeypair -storepass password -keyalg RSA -keysize 2048 -dname "CN=client" -alias client -ext "SAN:c=DNS:localhost,IP:127.0.0.1" -keystore quarkus-client-mtls/src/main/resources/META-INF/resources/client.keystore
+----
+
+NOTE: For this example, the `client.keystore` used as _truststore_ for the server application and in the same way the `server.keystore` for the client application.
+
+[source]
+----
+cp quarkus-server-mtls/src/main/resources/META-INF/resources/server.keystore quarkus-client-mtls/src/main/resources/META-INF/resources/client.truststore
+
+cp quarkus-client-mtls/src/main/resources/META-INF/resources/client.keystore quarkus-server-mtls/src/main/resources/META-INF/resources/server.truststore
+----
+
+== Hello Server Application.
+
+[source]
+----
+code quarkus-server-mtls/
+----
+
+=== Enable SSL On the Hello Server Application.
+
+Add the following properties to enable SSL in your application `src/main/resources/application.properties`
+
+.application.properties
+[source]
+----
+quarkus.ssl.native=true
+
+quarkus.http.ssl-port=8443
+quarkus.http.ssl.certificate.key-store-file=META-INF/resources/server.keystore
+quarkus.http.ssl.certificate.key-store-password=password
+
+quarkus.http.port=0
+quarkus.http.test-port=0 
+----
+
+NOTE: See the guide https://quarkus.io/guides/native-and-ssl[Using SSL with Native] to explore in details how SSL works in Quarkus.
+
+=== Activate client authentication.
+
+.application.properties
+[source]
+----
+quarkus.http.ssl.client-auth=required
+quarkus.http.ssl.certificate.trust-store-file=META-INF/resources/server.truststore
+quarkus.http.ssl.certificate.trust-store-password=password
+----
+
+=== Update GreetingResource
+
+To understand if the response is coming from the server application, you will update the `org.acme.server.mtls.GreetingResource` class.
+
+.org.acme.server.mtls.GreetingResource
+[source,java]
+----
+@Path("/hello")
+public class GreetingResource {
+
+    @GET
+    @Produces(MediaType.TEXT_PLAIN)
+    public String hello() {
+        return "hello from server"; // <1>
+    }
+}
+----
+<1> Change the return statement in _"hello from server"_
+
+And the test class.
+
+.org.acme.server.mtls.GreetingResourceTest
+[source,java]
+----
+@QuarkusTest
+public class GreetingResourceTest {
+
+    @Test
+    public void testHelloEndpoint() {
+        given()
+          .when().get("/hello")
+          .then()
+             .statusCode(200)
+             .body(is("hello from server")); <1>
+    }
+
+}
+----
+<1> Change the matcher in _"hello from server"_
+
+=== Run it.
+
+[source]
+----
+mvn quarkus:dev
+----
+
+If everything is correct when you try to hit the _/hello_ endpoint, you should expect the following error.
+
+[source]
+----
+curl -k https://localhost:8443
+curl: (35) error:1401E412:SSL routines:CONNECT_CR_FINISHED:sslv3 alert bad certificate
+----
```

Comment:
This means that your client (`curl`) did not provide a trusted certificate when it connected to the server.